doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,334 @@
+  tx: mpsc::Sender<(GlobalId, Vec<(Row, Diff)>)>,
+}
+```
+
+In addition to the `CollectionAppender`, we'll update the `CollectionManager` to receive a
+`oneshot::Sender<Result<(), StorageError>>` that it can use to optionally notify any listeners when
+and if the updates were successfully committed.
+
+Adding the `CollectionAppender` introduces a possible correctness issue though:
+
+1. Create a `CollectionAppender` for `GlobalId(A)`.
+2. Drop source associated with `GlobalId(A)`.
+3. Try to append data.
+
+To protect against this we can change the `CollectionManager` to return an error via the
+`oneshot::Sender` if the [provided GlobalId no longer exists](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs#L83).
+
+Adding this new API gets us two important properties that the existing `append` API does not have:
+
+1. The ability to send updates to the `StorageController` from a thread other than the main
+   `Coordinator` thread.
+2. Allows the `StorageController` to pick the timestamp for the update, instead of the
+   `Coordinator`, which aligns with our long term goals if having a storage owned "push source" in
+   the future.
+
+### Request Validation
+[request_validation]: #request_validation
+
+Every `WEBHOOK` source will be open to the entire internet, as such we need to include some sort of
+validation so we can verify that requests are legitimate. This is can be fairly difficult to do
+though because it could require us to run arbitrary logic when receiving each event, e.g. something
+very similar to `TRANSFORM USING`. Scanning a few applications there seem to be two common ways of
+validating webhook requests:
+
+1. HMAC SHA256 of the request body using a shared secret, with the signature being stored in a
+   header.
+2. Shared token being present in a header.
+
+As such, I believe we can introduce specific syntax to cover these two scenarios, eliminating the
+need to support running arbitrary logic when receiving an event. We can introduce new keywords,
+`VALIDATE USING ( ... )` which contains the statement used to validate a request. In the future
+this statement can be arbirary, the two predefined statements we'll support today are:
+
+```
+CREATE SECRET segment_hmac_key = "<SEGMENT_HMAC_KEY>";
+
+VALIDATE USING (
+  HMAC KEY = SECRET segment_hmac_key,
+  HMAC BODY SHA256 SIGNATURE HEADER = 'X-Signature',
+)
+```
+
+or
+
+```
+CREATE SECRET buildkite_webhook_token = "<BUILDKITE_WEBHOOK_TOKEN>";
+
+VALIDATE USING (
+  TOKEN = SECRET buildkite_webhook_token,
+  TOKEN EXISTS IN HEADER = 'X-Token',
+)
+```
+
+In the future the statement could be anything that returns a single bool, for example:
+
+```
+CREATE SECRET hmac_key = "<hmac_key>";
+
+VALIDATE USING (
+  SELECT signature = hmac('sha256', hmac_key, payload)
+  FROM (
+    SELECT
+      header->>'X-Signature' as signature,
+      base64('timestamp=' || header->>'timestamp' || 'body=' || body) as payload
+  )
+)
+```
```

Comment:
In the past we have had more luck when we've started with the general form and layered on the syntax sugar later, once we understand what the most common patterns are. Manual cluster replicas evolving into managed clusters is an example of us doing this well. The `FORMAT` and `ENVELOPE` clauses on `CREATE SOURCE` are an example of us doing this poorly; the complexity in the interaction between connections, formats, and envelopes repeatedly confuses users. We'd be much better off if we had started with `TRANSFORM USING` and only layered on `FORMAT` and `ENVELOPE` once we understood the most common transformations.

Taking a quick look, `HMAC KEY` and `HMAC BODY SHA256 SIGNATURE HEADERS` options don't seem sufficiently expressive to handle Segment/GitHub/Buildkite signatures. Validating Buildkite signatures requires this code...

```rb
class BuildkiteWebhook
  def self.valid?(webhook_request_body, header, secret)
    timestamp, signature = get_timestamp_and_signatures(header)
    expected = OpenSSL::HMAC.hexdigest("sha256", secret, "#{timestamp}.#{webhook_request_body}")
    Rack::Utils.secure_compare(expected, signature)
  end

  def self.get_timestamp_and_signatures(header)
    parts = header.split(",").map { |kv| kv.split("=", 2).map(&:strip) }.to_h
    [parts["timestamp"], parts["signature"]]
  end
end
```

(although you could alternatively use tokens with Buildkite, which are much simpler)

...while validating GitHub signatures requires this code...

```rb
def verify_signature(payload_body)
  signature = 'sha256=' + OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'), ENV['SECRET_TOKEN'], payload_body)
  return halt 500, "Signatures didn't match!" unless Rack::Utils.secure_compare(signature, request.env['HTTP_X_HUB_SIGNATURE_256'])
end
```

...and Segment's signature is like GitHub's but without the `sha256=` prefix.

As one more datapoint, [Stripe has an even more complicated scheme](https://stripe.com/docs/webhooks/signatures#verify-manually).

All this to me points to needing to allow arbitrary expressions for webhook validation, since each webhook provider has such a different take on what the right way to send a signature is.

I think my proposal would be that we support syntax like this:

```
VALIDATE USING (
    headers['X-Signature'] = hmac('sha256', SECRET webhook_secret, body)
)
```

Notes on semantics:

  * `body` is the raw body text _regardless_ of what `FORMAT` you've asked for. This is important when the HMAC needs to be calculated against the raw JSON, not the prettified JSON you might get from going from text -> json -> text.
  * `headers` is a `map[text => text]` containing the values of all headers in the request. It is available to the expression in `VALIDATE USING` even if `INCLUDE HEADERS` is not specified.
  * `SECRET` is special syntax, only valid in a `VALIDATE USING` clause, which names a secret whose value should be injected into the expression.
  * The expression must have type `boolean`.
  * ⚠️ there are some really weird questions around `bytea` vs `text` to sort out. HTTP bodies and headers can technically be bytes, not text, and so can secrets. Really not sure what do about that. Almost all HTTP requests are representable as text, so maybe we can just ignore bytes.

And then we'd provide a bunch of `VALIDATE USING` "recipes" for common webhook providers, but sufficiently sophisticated users could also write their own validation expressions.

Down the line, I can definitely imagine that we might want to make this more ergonomic by having options for common validations. But maybe that looks more like `VALIDATE STRIPE (SECRET ...)` or `VALIDATE BUILDKITE (TOKEN ...)` where we handle _all_ the details for you for common webhook providers! It's hard to say until we have a little more experience with what providers our users are commonly connecting Materialize to.

Reply:
This is super! Thank you for writing all of this out! I added "supporting common recipes" to the future work